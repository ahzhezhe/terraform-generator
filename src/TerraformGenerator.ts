import child_process from 'child_process';
import fs from 'fs';
import path from 'path';
import shell from 'shelljs';
import { Block, Comment, Resource, Data, Module, Output, Provider, Variable, Backend, Provisioner, ResourceToDataOptions, Locals, Import, ImportArgs, VariableArgs, ModuleArgs, OutputArgs, Moved, MovedArgs, RemovedArgs, Removed } from './blocks';
import { TerraformArgs, Util } from './utils';

/**
 * @category TerraformGenerator
 */
export interface WriteOptions {
  /**
   * Directory to write to.
   *
   * Default = .
   */
  dir?: string;
  /**
   * Terraform filename, must ends with .tf.
   *
   * Default = terraform.tf
   */
  tfFilename?: string;
  /**
   * Terraform variables filename, must ends with .tfvars.
   *
   * Default = terraform.tfvars
   */
  tfvarsFilename?: string;
  /**
   * Put `true` to use `terraform fmt` to format the configuration, Terraform must be installed.
   *
   * If the terraform binary is named differently in your machine, put the binary name instead.
   *
   * Default = false
   */
  format?: boolean | string;
}

/**
 * @category TerraformGenerator
 */
export class TerraformGenerator {

  readonly #arguments?: TerraformArgs;
  readonly #blocks: Block[] = [];
  #variables: TerraformArgs = {};

  /**
   * Construct Terraform generator.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args?: TerraformArgs) {
    this.#arguments = args;
  }

  /**
   * Generate Terraform configuration as string.
   */
  generate(): { tf: string; tfvars?: string } {
    return {
      tf: this.#generateTf(),
      tfvars: Object.keys(this.#variables).length > 0 ? this.#generateTfvars() : undefined
    };
  }

  #generateTf(): string {
    let str = '';

    if (this.#arguments || this.#blocks.filter(block => block.isInsideTerraformBlock()).length > 0) {
      str += 'terraform {\n';
      str += Util.argumentsToString(this.#arguments);
      this.#blocks.forEach(block => {
        if (block.isInsideTerraformBlock()) {
          str += block.toTerraform();
        }
      });
      str += '}\n\n';
    }

    this.#blocks.forEach(block => {
      if (!block.isInsideTerraformBlock()) {
        str += block.toTerraform();
      }
    });

    return Util.unescape(str);
  }

  #generateTfvars(): string {
    let str = '';

    Object.keys(this.#variables).forEach(key => {
      str += Util.argumentsToString({ [key]: this.#variables[key] });
      str += '\n';
    });

    return str;
  }

  /**
   * Write Terraform configuration to a file.
   *
   * @param options options
   */
  write(options?: WriteOptions): void {
    if (!options) {
      options = {};
    }
    if (!options.dir) {
      options.dir = '.';
    }
    if (!options.tfFilename) {
      options.tfFilename = 'terraform.tf';
    }
    if (!options.tfFilename.endsWith('.tf')) {
      options.tfFilename += '.tf';
    }
    if (!options.tfvarsFilename) {
      options.tfvarsFilename = 'terraform.tfvars';
    }
    if (!options.tfvarsFilename.endsWith('.tfvars')) {
      options.tfvarsFilename += '.tfvars';
    }
    if (options.format === true) {
      options.format = 'terraform';
    }

    const result = this.generate();
    shell.mkdir('-p', options.dir);
    fs.writeFileSync(path.join(options.dir, options.tfFilename), result.tf);
    if (result.tfvars) {
      fs.writeFileSync(path.join(options.dir, options.tfvarsFilename), result.tfvars);
    }

    if (options.format) {
      child_process.execSync(`${options.format} fmt`, { cwd: options.dir });
    }
  }

  /**
   * Add blocks into Terraform.
   *
   * @param blocks blocks
   */
  addBlocks(...blocks: Block[]): this {
    blocks.forEach(block => this.#blocks.push(block));
    return this;
  }

  /**
   * Add comment into Terraform.
   *
   * @param comment comment
   */
  comment(comment: string): Comment {
    const block = new Comment(comment);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add provider into Terraform.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  provider(type: string, args: TerraformArgs): Provider {
    const block = new Provider(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add resource into Terraform.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param name name
   * @param args arguments
   * @param provisioners provisioners
   */
  resource(type: string, name: string, args: TerraformArgs, provisioners?: Provisioner[]): Resource {
    const block = new Resource(type, name, args, provisioners);
    this.addBlocks(block);
    return block;
  }

  /**
   * Convert resource into data source and add it into Terraform.
   *
   * @param resource resource
   * @param options options
   * @param argNames names of resource arguments to be converted into data source arguments;
   * use array for name mapping, position 0 = original resource's argument name, position 1 = mapped data source's argument name
   * @param args extra arguments
   */
  dataFromResource(
    resource: Resource,
    options: ResourceToDataOptions | undefined,
    argNames: (string | [string, string])[],
    args?: TerraformArgs
  ): Data {
    const block = resource.toData(options, argNames, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add data source into Terraform.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param name name
   * @param args arguments
   */
  data(type: string, name: string, args: TerraformArgs): Data {
    const block = new Data(type, name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add module into Terraform.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   */
  module(name: string, args: ModuleArgs): Module {
    const block = new Module(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add output into Terraform.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   */
  output(name: string, args: OutputArgs): Output {
    const block = new Output(name, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add locals into Terraform.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  locals(args: TerraformArgs): Locals {
    const block = new Locals(args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add variable into Terraform.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param name name
   * @param args arguments
   * @param value variable value
   */
  variable(name: string, args: VariableArgs, value?: any): Variable {
    const block = new Variable(name, args);
    this.addBlocks(block);
    if (value != null) {
      this.addVars({ [name]: value });
    }
    return block;
  }

  /**
   * Add import into Terraform.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  import(args: ImportArgs): Import {
    const block = new Import(args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add backend into Terraform.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param type type
   * @param args arguments
   */
  backend(type: string, args: TerraformArgs): Backend {
    const block = new Backend(type, args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add moved into Terraform.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param args arguments
   */
  moved(args: MovedArgs): Moved {
    const block = new Moved(args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add removed into Terraform.
   *
   * Refer to Terraform documentation on what can be put as type & arguments.
   *
   * @param args arguments
   */
  removed(args: RemovedArgs): Removed {
    const block = new Removed(args);
    this.addBlocks(block);
    return block;
  }

  /**
   * Add variable values into Terraform.
   *
   * @param variables variables
   */
  addVars(variables: TerraformArgs): this {
    this.#variables = {
      ...this.#variables,
      ...variables
    };
    return this;
  }

  /**
   * Merge this instance with other TerraformGenerator instances.
   *
   * @param tfgs other instances
   */
  merge(...tfgs: TerraformGenerator[]): this {
    tfgs.forEach(tfg => {
      this.addBlocks(...tfg.getBlocks());
      this.addVars(tfg.getVars());
    });
    return this;
  }

  /**
   * Get arguments.
   */
  getArguments(): TerraformArgs | undefined {
    return this.#arguments;
  }

  /**
   * Get blocks.
   */
  getBlocks(): Block[] {
    return this.#blocks;
  }

  /**
   * Get variables.
   */
  getVars(): TerraformArgs {
    return this.#variables;
  }

}
