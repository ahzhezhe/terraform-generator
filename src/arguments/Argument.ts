import { TerraformElement, Util } from '../utils';

/**
 * @category Argument
 */
export class Argument<T extends string = string> extends TerraformElement {

  readonly #argument: T;

  /**
   * Construct argument.
   *
   * @param arg argument as string
   */
  constructor(arg: T) {
    super();
    this.#argument = arg;
  }

  /**
   * Get argument's attribute.
   *
   * @param name attribute name
   */
  attr(name: string): Argument {
    return new Argument(`${this.#argument}.${name.trim()}`);
  }

  /**
   * Get list argument's element.
   *
   * @param idx element index
   */
  element(idx: number): Argument {
    return new Argument(`${this.#argument}[${idx}]`);
  }

  /**
   * To Terraform representation.
   *
   * Use this method when argument is used as an interpolation in another Terraform argument or code.
   */
  override toTerraform(): string {
    return Util.escape(this.#argument);
  }

  /**
   * To string.
   *
   * Use this method when argument is used as an interpolation in a Terraform string or heredoc.
   *
   * It is automatically called when argument is used in template literal.
   */
  toString(): string {
    return `\${${this.toTerraform()}}`;
  }

}

/**
 * Convenient function to construct new [[Argument]].
 *
 * @param arg argument as string or copy from another argument object
 *
 * @category Argument
 */
export const arg = <T extends string = string>(arg: T): Argument<T> => new Argument(arg);
