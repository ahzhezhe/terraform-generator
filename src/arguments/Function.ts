import { Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Function extends Argument {

  /**
   * Construct function argument.
   *
   * @param name function name
   * @param args function arguments
   */
  constructor(name: string, ...args: any[]) {
    super(Function.#constructArgument(name, ...args));
  }

  static #constructArgument(fn: string, ...args: any[]): string {
    let str = `${fn}(`;
    args.forEach((arg, i) => {
      str += Util.argumentValueToString(arg);
      if (i < args.length - 1) {
        str += ', ';
      }
    });
    str += ')';
    return str;
  }

}

/**
 * Convenient function to construct new [[Function]].
 *
 * @param name function name
 * @param args function arguments
 *
 * @category Argument
 */
// eslint-disable-next-line @typescript-eslint/ban-types
export const fn = (name: string, ...args: any[]): Function => new Function(name, ...args);
