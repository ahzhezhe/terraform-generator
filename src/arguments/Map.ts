import { TerraformArgs, Util } from '../utils';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Map extends Argument {

  readonly arguments: TerraformArgs;

  /**
   * Construct map.
   *
   * @param args map values
   */
  constructor(args: TerraformArgs) {
    super(Util.argumentValueToString(args)!);
    this.arguments = args;
  }

}

/**
 * Convenient function to construct new [[Map]].
 *
 * @param args map values
 *
 * @category Argument
 */
export const map = (args: TerraformArgs): Map => new Map(args);
