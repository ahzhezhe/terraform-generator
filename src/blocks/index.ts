export * from './Block';
export * from './Backend';
export * from './Comment';
export * from './Data';
export * from './Import';
export * from './Locals';
export * from './Module';
export * from './Moved';
export * from './Output';
export * from './Provider';
export * from './Provisioner';
export * from './Removed';
export * from './Resource';
export * from './Variable';
