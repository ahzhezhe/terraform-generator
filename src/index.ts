export { TerraformElement, TerraformArgs } from './utils';

export { TerraformGenerator, WriteOptions } from './TerraformGenerator';

export {
  Block,
  Backend,
  Comment,
  Data,
  Import, ImportArgs,
  Locals,
  Module, ModuleArgs,
  Moved, MovedArgs,
  Output, OutputArgs,
  Provider,
  Provisioner, FileProvisionerArgs, LocalExecProvisionerArgs, RemoteExecProvisionerArgs,
  Removed, RemovedArgs,
  Resource, ResourceToDataOptions,
  Variable, VariableArgs
} from './blocks';

export {
  Argument, arg,
  Attribute, attr,
  Function, fn,
  Heredoc, heredoc,
  List, list,
  Map, map
} from './arguments';
