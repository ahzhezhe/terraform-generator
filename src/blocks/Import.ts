import { Argument, Attribute } from '../arguments';
import { TerraformElement, Util } from '../utils';
import { Block } from '.';

/**
 * @category Block
 */
export interface ImportArgs {
  to: TerraformElement;
  id: TerraformElement | string;
  provider?: Argument;
}

/**
 * @category Block
 */
export class Import extends Block<ImportArgs> {

  /**
   * Construct import.
   *
   * Refer to Terraform documentation on what can be put as arguments.
   *
   * @param args arguments
   */
  constructor(args: ImportArgs) {
    super('import', [], args);
  }

  override asArgument(): Argument {
    throw Util.inaccessibleMethod();
  }

  override attr(_name: string): Attribute {
    throw Util.inaccessibleMethod();
  }

}
