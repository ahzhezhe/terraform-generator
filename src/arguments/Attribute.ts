import { Block } from '../blocks';
import { Argument } from '.';

/**
 * @category Argument
 */
export class Attribute extends Argument {

  /**
   * Construct block's attribute.
   *
   * @param block block
   * @param attrName attribute name
   */
  constructor(block: Block, attrName: string) {
    super(`${block.asArgument().toTerraform()}.${attrName.trim()}`);
  }

}

/**
 * Convenient function to construct new block's [[Attribute]].
 *
 * @param block block
 * @param name attribute name
 *
 * @category Argument
 */
export const attr = (block: Block, name: string): Attribute => new Attribute(block, name);
