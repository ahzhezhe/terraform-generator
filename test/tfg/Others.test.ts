import fs from 'fs';
import path from 'path';
import { Argument, arg, map } from '../../src/arguments';
import { Provisioner } from '../../src/blocks';
import { TerraformGenerator } from '../../src/TerraformGenerator';

const createTerraformGenerator = (): TerraformGenerator => {
  const tfg = new TerraformGenerator();

  tfg.variable('test', {
    type: arg('string')
  });

  tfg.variable('test2', {
    type: arg('string')
  }, 'test');

  tfg.data('aws_vpc', 'test', {
    cidr_block: 'test'
  });

  tfg.module('test', {
    source: './test'
  });

  const r = tfg.resource('aws_vpc', 'test', {
    cidr_block: 'test',
    tags: map({
      a: 'a'
    })
  });

  tfg.dataFromResource(r, undefined, ['cidr_block', ['tags', 'tag']]);
  tfg.dataFromResource(r, { name: 'test2' }, ['cidr_block', ['tags', 'tag']]);

  const resource = tfg.resource('innerBlock', 'innerBlock', {
    a: 'a'
  }, [
    new Provisioner('local-exec', {
      command: 'echo hello'
    }),
    new Provisioner('local-exec', {
      command: 'echo world'
    })
  ]);

  const locals = tfg.locals({
    a: 'a',
    b: 123,
    c: r.attr('x')
  });
  tfg.resource('locals', 'locals', {
    a: locals.arg('a')
  });

  tfg.import({
    to: resource,
    id: 'id',
    provider: arg('arg')
  });

  tfg.resource('tags', 'tags', {
    tags: map({
      'a': 'a',
      'b': 'b c d'
    })
  });

  tfg.comment('comment');

  tfg.comment(`
    line1
    line2
    line3
    line4
  `);

  tfg.moved({
    from: new Argument('resource.a'),
    to: resource
  });

  tfg.removed({
    from: new Argument('resource.b'),
    lifecycle: {
      destroy: false
    }
  });

  const tfg2 = new TerraformGenerator();
  tfg2.resource('tfg2', 'tfg2', {
    tfg2: 'tfg2'
  });

  tfg.merge(tfg2);

  return tfg;
};

const outputDir = path.join('test', '__output__');

test('Others', () => {
  const tfg = createTerraformGenerator();

  expect(tfg.getArguments()).toMatchSnapshot();
  expect(tfg.getBlocks()).toMatchSnapshot();
  expect(tfg.getVars()).toMatchSnapshot();
  expect(tfg.generate()).toMatchSnapshot();

  tfg.write({
    dir: outputDir,
    tfFilename: 'others',
    tfvarsFilename: 'others'
  });
  const tf = fs.readFileSync(path.join(outputDir, 'others.tf'), 'utf8');
  expect(tf).toMatchSnapshot();
  const tfvars = fs.readFileSync(path.join(outputDir, 'others.tfvars'), 'utf8');
  expect(tfvars).toMatchSnapshot();
});
