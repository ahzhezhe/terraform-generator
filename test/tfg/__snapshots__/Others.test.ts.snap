// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Others 1`] = `undefined`;

exports[`Others 2`] = `
[
  Variable {
    "blockNames": [
      "test",
    ],
    "blockType": "variable",
    "name": "test",
  },
  Variable {
    "blockNames": [
      "test2",
    ],
    "blockType": "variable",
    "name": "test2",
  },
  Data {
    "blockNames": [
      "aws_vpc",
      "test",
    ],
    "blockType": "data",
    "name": "test",
    "type": "aws_vpc",
  },
  Module {
    "blockNames": [
      "test",
    ],
    "blockType": "module",
    "name": "test",
  },
  Resource {
    "blockNames": [
      "aws_vpc",
      "test",
    ],
    "blockType": "resource",
    "name": "test",
    "type": "aws_vpc",
  },
  Data {
    "blockNames": [
      "aws_vpc",
      "test",
    ],
    "blockType": "data",
    "name": "test",
    "type": "aws_vpc",
  },
  Data {
    "blockNames": [
      "aws_vpc",
      "test2",
    ],
    "blockType": "data",
    "name": "test2",
    "type": "aws_vpc",
  },
  Resource {
    "blockNames": [
      "innerBlock",
      "innerBlock",
    ],
    "blockType": "resource",
    "name": "innerBlock",
    "type": "innerBlock",
  },
  Locals {
    "blockNames": [],
    "blockType": "locals",
  },
  Resource {
    "blockNames": [
      "locals",
      "locals",
    ],
    "blockType": "resource",
    "name": "locals",
    "type": "locals",
  },
  Import {
    "blockNames": [],
    "blockType": "import",
  },
  Resource {
    "blockNames": [
      "tags",
      "tags",
    ],
    "blockType": "resource",
    "name": "tags",
    "type": "tags",
  },
  Comment {
    "blockNames": [],
    "blockType": "comment",
    "comment": "comment",
  },
  Comment {
    "blockNames": [],
    "blockType": "comment",
    "comment": "
    line1
    line2
    line3
    line4
  ",
  },
  Moved {
    "blockNames": [],
    "blockType": "moved",
  },
  Removed {
    "blockNames": [],
    "blockType": "removed",
  },
  Resource {
    "blockNames": [
      "tfg2",
      "tfg2",
    ],
    "blockType": "resource",
    "name": "tfg2",
    "type": "tfg2",
  },
]
`;

exports[`Others 3`] = `
{
  "test2": "test",
}
`;

exports[`Others 4`] = `
{
  "tf": "variable "test"{
type = string
}

variable "test2"{
type = string
}

data "aws_vpc" "test"{
cidr_block = "test"
}

module "test"{
source = "./test"
}

resource "aws_vpc" "test"{
cidr_block = "test"
tags = {
a = "a"
}
}

data "aws_vpc" "test"{
filter {
name = "tag:a"
values = [
"a"
]
}
cidr_block = "test"
}

data "aws_vpc" "test2"{
filter {
name = "tag:a"
values = [
"a"
]
}
cidr_block = "test"
}

resource "innerBlock" "innerBlock"{
a = "a"
provisioner "local-exec"{
command = "echo hello"
}
provisioner "local-exec"{
command = "echo world"
}
}

locals{
a = "a"
b = 123
c = aws_vpc.test.x
}

resource "locals" "locals"{
a = local.a
}

import{
to = innerBlock.innerBlock
id = "id"
provider = arg
}

resource "tags" "tags"{
tags = {
a = "a"
b = "b c d"
}
}

# comment

# line1
# line2
# line3
# line4

moved{
from = resource.a
to = innerBlock.innerBlock
}

removed{
from = resource.b
lifecycle {
destroy = false
}
}

resource "tfg2" "tfg2"{
tfg2 = "tfg2"
}

",
  "tfvars": "test2 = "test"

",
}
`;

exports[`Others 5`] = `
"variable "test"{
type = string
}

variable "test2"{
type = string
}

data "aws_vpc" "test"{
cidr_block = "test"
}

module "test"{
source = "./test"
}

resource "aws_vpc" "test"{
cidr_block = "test"
tags = {
a = "a"
}
}

data "aws_vpc" "test"{
filter {
name = "tag:a"
values = [
"a"
]
}
cidr_block = "test"
}

data "aws_vpc" "test2"{
filter {
name = "tag:a"
values = [
"a"
]
}
cidr_block = "test"
}

resource "innerBlock" "innerBlock"{
a = "a"
provisioner "local-exec"{
command = "echo hello"
}
provisioner "local-exec"{
command = "echo world"
}
}

locals{
a = "a"
b = 123
c = aws_vpc.test.x
}

resource "locals" "locals"{
a = local.a
}

import{
to = innerBlock.innerBlock
id = "id"
provider = arg
}

resource "tags" "tags"{
tags = {
a = "a"
b = "b c d"
}
}

# comment

# line1
# line2
# line3
# line4

moved{
from = resource.a
to = innerBlock.innerBlock
}

removed{
from = resource.b
lifecycle {
destroy = false
}
}

resource "tfg2" "tfg2"{
tfg2 = "tfg2"
}

"
`;

exports[`Others 6`] = `
"test2 = "test"

"
`;
